Use Tuna for tuning in rocMLIR weekly and nightly.

pcf000 opened the pull request:
Add --tflops option to include that in the .tsv file for Tuna.
Rearrange options slightly in conv-configs.
Initial unified Tuna script for CI.
Add export of results.
Running mysqld in user mode.
Quiet some logging.  Adjust tuna-script.sh to fit into CI better. Ready or not, here come the Tuna versions of the main tuning runs. Add toCommandLine methods to canonicalise configs to look up in tuningDb. Add multi_gpu to gfx908 labels to avoid single-GPU host. Improved local-mysql startup in tuna-script.sh, given libaio1 in the docker image. Make sure PATH and TUNA_DB_HOSTNAME are set when mysqld is already up. Quick-tuning support in the Tuna script.
Back to a development branch for quick-tuning and backward-data support. Tuning-space and load-factor support in tuna-script.sh. Results-to-stdout support in tuningRunner.py.
Use a temp directory instead of separate filenames to avoid rocprof stats collisions. More robust chdir for verify, support for attention op in tuna-script.sh. And turn on Tuna attention support in Jenkinsfile.

pcf000 committed:
Add --tflops option to include that in the .tsv file for Tuna.
Rearrange options slightly in conv-configs.
Initial unified Tuna script for CI.
Add export of results.
Running mysqld in user mode.
Quiet some logging.  Adjust tuna-script.sh to fit into CI better.
Ready or not, here come the Tuna versions of the main tuning runs.
Add toCommandLine methods to canonicalise configs to look up in tuningDb.
Add multi_gpu to gfx908 labels to avoid single-GPU host.
Improved local-mysql startup in tuna-script.sh, given libaio1 in the docker image.
Make sure PATH and TUNA_DB_HOSTNAME are set when mysqld is already up.
Quick-tuning support in the Tuna script.
Back to a development branch for quick-tuning and backward-data support.
Tuning-space and load-factor support in tuna-script.sh.
Results-to-stdout support in tuningRunner.py.
Use a temp directory instead of separate filenames to avoid rocprof stats collisions.
More robust chdir for verify, support for attention op in tuna-script.sh.
And turn on Tuna attention support in Jenkinsfile.

pcf000 commented on mlir/tools/rocmlir-gen/rocmlir-gen.cpp:
```diff
@@ -245,17 +245,17 @@ static llvm::cl::opt<int>
 
 /// Matrix options
 static llvm::cl::opt<int64_t> gemmM("m",
-                                    llvm::cl::desc("M dimennsion of gemm()"),
```
Typos.

pcf000 commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -198,7 +198,7 @@ String getLabelFromChip(String chip) {
             return getLabelFromCodepath("vanilla")
         case "gfx908":
-            return "mlir && gfx908"
+            return "mlir && gfx908 && multi_gpu"
```
Tuna's chief benefit is parallelism.  We have four gfx908s in the rocmhead pool, with 6, 4, 4, and 1 GPU respectively;  don't use the 1-GPU one.

pcf000 commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -546,25 +546,25 @@ pipeline {
-                                    sh """python3 ./bin/tuningRunner.py --operation gemm \
-                                            --configs_file=../mlir/utils/jenkins/ci-configs/selected-gemm-configs \
-                                            --output=tuning_gemm.tsv
+                                    sh """../mlir/utils/tuna/tuna-script.sh -o gemm \
```
Replace invocation of tuningRunner.py with use of tuna-script.sh, which will set up for Tuna and then do the same tuning.

pcf000 commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -546,25 +546,25 @@ pipeline {
-                                    sh """python3 ./bin/tuningRunner.py --operation attention \
-                                        --configs_file=../mlir/utils/jenkins/ci-configs/selected-attention-configs \
-                                        --output=tuning_attention.tsv --verify-mode none
-                                        [ -f tuning_conv.tsv ]"""
```
Existing attention tuning was checking for wrong file, hence the longer diff.

pcf000 commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -712,36 +712,29 @@ pipeline {
-                                sh """python3 ./bin/tuningRunner.py --operation gemm \
-                                        --configs_file=../mlir/utils/performance/gemm-configs \
-                                        --output=mlir_tuning_${ARCH}.tsv
+                                // (Includes int8xint8->int8 for performance comparisons against CK.)
```
In other words, this one tuna-script.sh invocation accounts for both tuningRunner.py invocations at this point.  I chose to have configuration-generation make all the TxT->T cases, and special-cased i8 to make both i8xi8->i8 and i8xi8->i32.

pcf000 commented on mlir/utils/performance/conv-configs:
```diff
@@ -1,43 +1,43 @@
 # This section of the file comes from resnet50-miopen-configs
--n 256 -c 1024 -H 14 -W 14 -k 2048 -y 1 -x 1 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -t 1
```
Swapped "-m conv" and "-g 1" when I was trying to make all the input and output configs match, since perfRunner.py did a simple string-compare.  Eventually made toCommandLine and canonicalised them all at runtime.

pcf000 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -170,15 +170,18 @@ def read_tuning_db(path: Optional[str]) -> MaybeTuningDb:
+                elif len(entries) == 4:
                     arch, _, config, perfConfig = entries
                     ret[arch, config] = perfConfig
+                # note: 5-entry form includes tflops at end
```
I wanted TFlops in the tuningRunner.py output so I could include it in reports.

pcf000 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -420,6 +426,16 @@ def fromCommandLine(cls, argv, arch, numCU):
             convStrideH, convStrideW, paddingH, paddingW, dilationH, dilationW,
                    group, arch, numCU)
 
+    def toCommandLine(self):
```
Canonicalise so we can reliably compare them.

pcf000 commented on mlir/utils/performance/tuningRunner.py:
```diff
@@ -46,10 +49,8 @@ def verifyModeFlags(verifyMode: str) -> str:
 #Run a gemm or conv config and verify it
 def verifyKernelWithPerfConfig(perfConfig, config, paths: Paths, options: Options) -> float:
-    print(f"Verifying with perfConfig = {perfConfig}")
+    print(f"Verifying with perfConfig = {perfConfig}", file=sys.stderr)
```
Changed all the running-commentary prints to print to stderr, so I could optionally send the results to stdout.

pcf000 commented on mlir/utils/performance/tuningRunner.py:
```diff
@@ -46,10 +49,8 @@ def verifyModeFlags(verifyMode: str) -> str:
-    print(f"Verifying with perfConfig = {perfConfig}")
+    print(f"Verifying with perfConfig = {perfConfig}", file=sys.stderr)
     config.setPerfConfig(perfConfig.strip())
-    # remove the result file generated by rocprof in previous benchmarking
```
The next few changes have the verify execution write its results to a temporary directory, which will automatically go away at the end, to ensure that multiple threads don't step on each others' output.  We don't need to delete the last run's output here, because it already happened.

pcf000 commented on mlir/utils/performance/tuningRunner.py:
```diff
@@ -59,42 +60,48 @@ def verifyKernelWithPerfConfig(perfConfig, config, paths: Paths, options: Option
+        print(rocmlirGenCommand, file=sys.stderr)
+
+    prevdir = os.getcwd()
+    with tempfile.TemporaryDirectory() as tmpdir:
```
Make a temp directory, cd to it, execute using the usual filename, extract the nanoSeconds.  The 'with' removes the directory, the 'finally' ensures that we always cd back where we came from.

krzysz00 replied:
Makes sense - I assume this auto-deletes after?

pcf000 commented on mlir/utils/performance/tuningRunner.py:
```diff
@@ -329,15 +352,27 @@ def main(args=None):
-        for testVector, perfConfig in winners.items():
-            print(f"Arch = {arch}({numCU} CUs), vector = '{testVector}', perfConfig = {perfConfig}")
-            print(f"{arch}\t{numCU}\t{testVector}\t{perfConfig}", file=outFile)
+    if parsed_args.output == '-':
```
Writing results to stdout is another way to avoid collisions when multithreaded -- there's no file, thus no chance of identical filenames.

pcf000 commented on mlir/utils/tuna/tuna-script.sh:
```diff
@@ -0,0 +1,155 @@
+#!/bin/bash
```
New file, to manage the Tuna sequence.

A stable persistent Tuna setup has a database running, Tuna installed, and database tables initialised.  At the moment, we don't have a persistent database, and thus have to do all the setup on each run.

Each run then imports configs, starts a session, sets up jobs, executes those jobs, and exports the results to a .tsv file for later perfRunner.py use.

pcf000 commented on mlir/utils/tuna/tuna-script.sh:
```diff
@@ -0,0 +1,155 @@
+function mysql_setup_generic
+{
+    # Latest "generic" version.  (Must have libaio1 installed already.)
+    # Note that all this happens without privileges.
```
rocMLIR CI jobs run as a normal user.  Embedding mysqld in our docker image is (a) large and (b) painful.  The pain comes from not having systemd in docker-world and then needing root to start things up.  Instead, we install the "generic" mysql, which doesn't care about systemd, and then arrange to run it as a normal user, in part by using a socket instead of dealing with network ports and authorisation.

pcf000 commented on mlir/utils/tuna/tuna-script.sh:
```diff
@@ -0,0 +1,155 @@
+{
+    startdir=`pwd`
+    rm -rf MITuna
+    git clone --branch pf-tuna-rocmlir-3 https://github.com/ROCmSoftwarePlatform/MITuna.git
```
Clone Tuna and do python things.  My development branch right now, rather than wait for the Tuna pull request.  A lot of python modules are installed, but at least it only happens once and they stick around for subsequent tuna-script.sh invocations in the same docker container.  Add our database tables at the end.

krzysz00 replied:
Can we move this into the Jenkinsfile so that it picks up on any weird config that Jenkins does?

pcf000 replied:
I hesitate, because then any desired Tuna update will require a docker image re-generation.

pcf000 commented on mlir/utils/tuna/tuna-script.sh:
```diff
@@ -0,0 +1,155 @@
+{
+    tablekind=$1
+
+    if [ "$tablekind" = "convolution" ]; then
```
I was a little sloppy about naming.  The other ops only have one variation.

krzysz00 replied:
`[[` here if possible

... heck, run `shellcheck` on this script just to be safe

krzysz00 replied:
Re-raise nitpick / request for a `shellcheck`

pcf000 replied:
Ran shellcheck, and it didn't care about [ vs [[, though it did point out some useful places to double-quote and also urged $() over ``.

pcf000 commented on mlir/utils/tuna/tuna-script.sh:
```diff
@@ -0,0 +1,155 @@
+{
+    kind=$1
+    space=$2
+    baselabel=`date --iso-8601=minutes`
```
Do the work.  Unfortunately, the tables are differentiated by config-type, so we have separate tables and separate jobs and separate runs for conv, gemm, and attention.  (Maybe we fix that later.)  Label starts with a timestamp to know when these ran.  Configs are global, but other parts are held under the umbrella of a "session" which includes the config-type and tuning-space;  jobs belong to a session and results are looked up by session.  Load-factor is an idea to explore how much parallelism we get and how much non-GPU overhead there is in a tuning run.

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -911,7 +932,8 @@ def runConfigWithMLIR(config: PerfConfiguration, paths: Paths, rocmlir_gen_flags
 def benchmarkMLIR(commandLine, confClass, paths: Paths, arch, numCU, tuningDb: MaybeTuningDb, rocmlir_gen_flags):
     config = confClass.fromCommandLine(commandLine, arch, numCU)
-    configStr = ' '.join(commandLine)
+    #configStr = ' '.join(commandLine)
```
Don't leave stray commented code in here?

krzysz00 commented on mlir/utils/performance/tuningRunner.py:
```diff
@@ -59,42 +60,48 @@ def verifyKernelWithPerfConfig(perfConfig, config, paths: Paths, options: Option
         result = result.decode('utf-8').strip()
-        if i > 0 and i % 100 == 0:
-            print(f"Tested {i} configs, best perf {maxTFlops} TFlops on perf_config {winningConfig}")
+        if not options.quiet and i > 0 and i % 100 == 0:
```
One note: Jenkins relies on scripts printing out to stderr/stdout regularly to enforce timeouts - are you sure you're not breaking this here?

krzysz00 commented on mlir/utils/performance/tuningRunner.py:
```diff
@@ -329,15 +352,27 @@ def main(args=None):
+    else:
+        outFile = open(parsed_args.output, 'a')
+
+    with outFile:
```
Does this work?

krzysz00 commented on mlir/utils/tuna/tuna-script.sh:
```diff
@@ -0,0 +1,155 @@
+    # Latest "generic" version.  (Must have libaio1 installed already.)
+    # Note that all this happens without privileges.
+    pushd /tmp
+    wget https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-8.0.34-linux-glibc2.28-x86_64.tar.gz
```
... all this goes in the Dockerfile, perhaps?

krzysz00 commented on mlir/utils/tuna/tuna-script.sh:
```diff
@@ -0,0 +1,155 @@
+    tar xf mysql-8.0.34-linux-glibc2.28-x86_64.tar.gz
+    mv mysql-8.0.34-linux-glibc2.28-x86_64 mysql
+    export PATH=$PATH:/tmp/mysql/bin
+    mysqld --initialize-insecure --datadir=/tmp/mysql/data
```
More accurately, I'd say the wget should be in the Dockerfile, and then we should create the database, either under /tmp/ or, more interestingly, under $WORKSPACE.

krzysz00 commented on mlir/utils/tuna/tuna-script.sh:
```diff
@@ -0,0 +1,155 @@
+    source myvenv/bin/activate
+
+    # --ignore-installed because of problems upgrading PyYAML.  See also -U.
+    python3 -m pip install -r requirements.txt --ignore-installed
```
... See above: why not do this in the Dockerfile?

krzysz00 reviewed (commented):
I'm not sure I like how much setup work is leaking into the actual CI run process here.

pcf000 committed:
Moving some setup from tuna-script.sh to Dockerfile (tuna-script.sh part)

pcf000 commented:
> I'm not sure I like how much setup work is leaking into the actual CI run process here.

Same here, but I was trying to get things in faster.  Anyway, I split the installation parts of setup out of tuna-script.sh and added them as final steps in Dockerfile;  the latter part is a separate pull request, because it'll have to be merged before this pull request will pass the tuning parts.

pcf000 committed:
Clone MITuna in tuna_setup, forcing the directory, and use /tuna-venv from Dockerfile.

pcf000 committed:
shellcheck:  double-quotes and backquotes

krzysz00 reviewed (approved):
Looks good to me

pcf000 committed:
Fix some items found in testing.

pcf000 committed:
Fix default TUNA_DIR to match what tuna_setup does.

pcf000 committed:
Troubleshooting potential lockhart proxy problem.

pcf000 committed:
Troubleshooting potential lockhart proxy problem.

pcf000 committed:
Troubleshooting potential lockhart proxy problem.

pcf000 committed:
Troubleshooting potential lockhart proxy problem.

pcf000 commented:
Reviewers, please let me do the merge.  All the checks may pass for bogus reasons while I work out the clone-MITuna details.

pcf000 committed:
Use Jenkins git instead of within tuna-script.sh.

pcf000 committed:
Using Jenkins git.

pcf000 committed:
Trying to get MITuna into a predictable place.

pcf000 committed:
Trying to get MITuna into a predictable place.

pcf000 committed:
Trying to get MITuna into a predictable place.

pcf000 committed:
Nested dirs.

pcf000 committed:
More directory manipulation.

pcf000 committed:
More directory manipulation.

pcf000 committed:
Trying direct sh.

pcf000 committed:
Trying direct sh.

pcf000 committed:
Trying direct sh.

pcf000 committed:
Now set http_proxy for lockhart.

pcf000 committed:
In both places.
